include/swift/ABI/TaskStatus.h

```diff
@@ -176,15 +181,18 @@ class TaskGroupTaskStatusRecord : public TaskStatusRecord {
 
   void detachChild(AsyncTask *child) {
     assert(child && "cannot remove a null child from group");
-    if (FirstChild == child) {
```

Comment:
Seems that would have been our offending not locked access AFAICS?

Reply:
This one is OK, the task status record lock is held when calling this. The non-locked access was:

```
  static AsyncTask *getNextChildTask(AsyncTask *task) {
    return task->childFragment()->getNextChild();
  }
```

Which gets called from `cancelAll`. We can run this, get the pointer to the next child, then another thread can `detachChild` and then free the task. If that happens before this thread uses the child pointer, it will end up with a dangling pointer.

Reply:
I see, thanks for the details!